contracts/upgradeable_contracts/amb_erc20_to_native/ForeignAMBErc20ToNative.sol

```diff
@@ -0,0 +1,174 @@
+    * @param _from address that will transfer the tokens to be locked.
+    * @param _receiver address that will receive the native tokens on the other network.
+    * @param _value amount of tokens to be transferred to the other network.
+    */
+    function relayTokens(address _from, address _receiver, uint256 _value) external {
+        require(_from == msg.sender || _from == _receiver);
+        _relayTokens(_from, _receiver, _value);
+    }
+
+    /**
+    * @dev It will initiate the bridge operation that will lock the amount of tokens transferred and mint the native tokens on
+    * the other network. The user should first call Approve method of the ERC677 token.
+    * @param _receiver address that will receive the native tokens on the other network.
+    * @param _value amount of tokens to be transferred to the other network.
+    */
+    function relayTokens(address _receiver, uint256 _value) external {
+        _relayTokens(msg.sender, _receiver, _value);
+    }
+
+    /**
+    * @dev Validates that the token amount is inside the limits, calls transferFrom to transfer the tokens to the contract
+    * and invokes the method to lock the tokens and mint the native tokens on the other network.
+    * The user should first call Approve method of the ERC677 token.
+    * @param _from address that will transfer the tokens to be locked.
+    * @param _receiver address that will receive the native tokens on the other network.
+    * @param _value amount of tokens to be transferred to the other network.
+    */
+    function _relayTokens(address _from, address _receiver, uint256 _value) internal {
+        // This lock is to prevent calling passMessage twice.
+        // When transferFrom is called, after the transfer, the ERC677 token will call onTokenTransfer from this contract
+        // which will call passMessage.
+        require(!lock());
+        ERC677 token = erc677token();
+        address to = address(this);
+        require(withinLimit(_value));
+        setTotalSpentPerDay(getCurrentDay(), totalSpentPerDay(getCurrentDay()).add(_value));
+
+        setLock(true);
+        token.transferFrom(_from, to, _value);
+        setLock(false);
+        bridgeSpecificActionsOnTokenTransfer(token, _from, _value, abi.encodePacked(_receiver));
+    }
+
+    /**
+    * @dev This method is called when transferAndCall is used from ERC677 to transfer the tokens to this contract.
+    * It will initiate the bridge operation that will lock the amount of tokens transferred and mint the native tokens on
+    * the other network.
+    * @param _from address that transferred the tokens.
+    * @param _value amount of tokens transferred.
+    * @param _data this parameter could contain the address of an alternative receiver of the tokens on the other network,
+    * otherwise it will be empty.
+    */
+    function onTokenTransfer(address _from, uint256 _value, bytes _data) external returns (bool) {
```

Comment:
Thought that probably it is OK that our mediators support ERC677 but then noticed that public availability of `erc677token` method could confuse users. Consider to introduce a new method `erc20token` that will return the value of `erc677token` (if it is possible to hide `erc677token` in this case at all?).

![image](https://user-images.githubusercontent.com/20793260/85841653-181a0000-b7a7-11ea-842e-9195c2bdcf9c.png)

Reply:
Agree. I hided the `erc677token()` method in erc-to-native and native-to-erc mediators, and introduced another one: `erc20token()`. Please, take a look.

Reply:
It is not correct to hide it for the `native-to-erc` mediators since they definitely work with a ERC677 token.